### PR TITLE
Hide features (PR to migrated Trac ticket #34185)

### LINKED ISSUE #34185: New doctest option to hide features
<div id="comment:0"></div>

As suggested in #33823 (see [comment 25](https://github.com/sagemath/sage/issues/33823#comment:25)) I make a first attempt for such an option here. The purpose of the option is to find test which need to be marked as `optional` but aren't right now.

For example see the test of the following file which contains two tests involving the optional package `database_knotinfo`. Note that on the system where I produced the output below this feature is available.

```bash
cat test_knotinfo.py
r"""
sage: KnotInfo.K9_10.homfly_polynomial()
-v^10*z^2 + v^8*z^4 - 2*v^10 + 2*v^8*z^2 + 2*v^6*z^4 + v^8 + 5*v^6*z^2 + v^4*z^4 + 2*v^6 + 2*v^4*z^2
sage: KnotInfo.K9_10.alexander_polynomial()  # optional database_knotinfo
4*t^4 - 8*t^3 + 9*t^2 - 8*t + 4
"""
```

The first test is missing the `optional` tag. Thus, this would fail on a system without that feature. But, running `sage -t` without the new option does not show the problem:

```bash
sage -t test_knotinfo.py
too many failed tests, not using stored timings
Running doctests with ID ...
...
Git ref: 9.7.beta5-...
...
Using --optional=database_knotinfo,debian,debugpy,mathics,mathics_scanner,palettable,pint,pip,sage,sage_spkg
Features to be detected: ...
Doctesting 1 file.
sage -t --random-seed=17350038375103993251556014960108337039 test_knotinfo.py
    [2 tests, 4.66 s]
----------------------------------------------------------------------
All tests passed!
----------------------------------------------------------------------
Total time for all tests: 4.7 seconds
    cpu time: 4.7 seconds
    cumulative wall time: 4.7 seconds
Features detected for doctesting:
```

Using the new option this will be the case:

```bash
sage -t --hide=database_knotinfo test_knotinfo.py
too many failed tests, not using stored timings
Running doctests with ...
...
Git ref: 9.7.beta5-...
...
Using --optional=debian,debugpy,mathics,mathics_scanner,palettable,pint,pip,sage,sage_spkg
Features to be detected: ...
Hidden features: database_knotinfo
Doctesting 1 file.
sage -t --random-seed=109756173219998976577146490280531589198 test_knotinfo.py
**********************************************************************
File "test_knotinfo.py", line 2, in test_knotinfo
Failed example:
    KnotInfo.K9_10.homfly_polynomial()
Exception raised:
    Traceback (most recent call last):
    ...
    NameError: name 'KnotInfo' is not defined
**********************************************************************
1 item had failures:
   1 of   2 in test_knotinfo
    [1 test, 1 failure, 0.00 s]
----------------------------------------------------------------------
sage -t --random-seed=109756173219998976577146490280531589198 test_knotinfo.py  # 1 doctest failed
----------------------------------------------------------------------
Total time for all tests: 0.0 seconds
    cpu time: 0.0 seconds
    cumulative wall time: 0.0 seconds
Features detected for doctesting:
```

Depends on #34282
Depends on #34283

CC:  @seblabbe

Component: **doctest framework**

Author: **Sebastian Oehms**

Branch/Commit: **[u/soehms/hide_features_34185](https://github.com/sagemath/sagetrac-mirror/tree/u/soehms/hide_features_34185) @ [`2a582f4`](https://github.com/sagemath/sagetrac-mirror/commit/2a582f47388352bbeea35c14168f21c2e01c4d3b)**

_Issue created by migration from https://trac.sagemath.org/ticket/34185_

### PR DESCRIPTION
**Pull request by soehms**
### :books: Description

See the description of the underlying migrated Trac ticket. This fixes #34185. 

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/features/__init__.py [hunk 1]
```diff
@@ -264,6 +293,109 @@ def resolution(self):
...
+            sage: from sage.features.graphviz import Graphviz
+            sage: Graphviz().joined_features()
+            [Feature('dot'), Feature('neato'), Feature('twopi')]
+            sage: from sage.features.interfaces import Mathematica
+            sage: Mathematica().joined_features()
+            []
+        """
+        from sage.features.join_feature import JoinFeature
+        if isinstance(self, JoinFeature):
+            return self._features
```
Comment (mkoeppe) [inline src/sage/features/__init__.py]: Perhaps this should yield recursively?

File: src/sage/features/__init__.py [hunk 2]
```diff
@@ -264,6 +293,109 @@ def resolution(self):
         self._cache_resolution = "\n".join(lines)
         return self._cache_resolution
 
+    def joined_features(self):
+        r"""
+        Return a list of features joined with ``self``.
```
Comment (mkoeppe) [inline src/sage/features/__init__.py]: perhaps "features that ``self`` is the join of"

File: src/bin/sage-runtests [hunk 3]
```diff
@@ -52,6 +52,10 @@ if __name__ == "__main__":
              'if set to "all", then all tests will be run; '
              'use "!FEATURE" to disable tests marked "# optional - FEATURE". '
              'Note that "!" needs to be quoted or escaped in the shell.')
+    parser.add_argument("--hide", metavar="FEATURES", default="",
+        help='run tests pretending that the software listed in FEATURES (separated by commas) is not installed; '
+             'if "all" is listed, will also hide features corresponding to all non standard packages; '
```
Comment (mkoeppe) [inline src/bin/sage-runtests]: I don't know what "non standard" means here; do you mean "optional or experimental"?

File: src/sage/features/__init__.py (hunk 1 above)
Comment (soehms) [reply src/sage/features/__init__.py]: Done!

File: src/sage/features/__init__.py (hunk 2 above)
Comment (soehms) [reply src/sage/features/__init__.py]: Done!

File: src/bin/sage-runtests (hunk 3 above)
Comment (soehms) [reply src/bin/sage-runtests]: > do you mean "optional or experimental"?

Yes! I changed this, to have it more clear!

**Comment by mkoeppe** [comment]
linter failure

**Review by mkoeppe — APPROVED** [review]

**Comment by soehms** [comment]
@mkoeppe Many thanks!

**Comment by kiwifb** [comment]
I'll  be honest, I only noticed this ticket because it is in Volker's merge queue. At first I thought it was a missed opportunity to drop some usage of stuff in `sage.misc.package` in doctest control and then to my horror, I noticed that it created new functionality in it and spread it tentacles in an area that was blissfully free of it before: "features". In fact, I took features as way to do things without `sage.misc.packages`. Well, no more!

`sage.misc.packages` is clearly an interface to consult the sage distribution packaging system. Its inclusion, use and any extensions goes against the effort of making sage more modular and pip installable as far as I am concerned. sage should be able to run independently of sage the distribution. That's the number one thing I would like to see split as a completely separate package and stopped being relied on for testing or anything else. This PR certainly hurt sage-on-distro as far as I am concerned.

**Comment by soehms** [comment]
> I'll be honest, I only noticed this ticket because it is in Volker's merge queue. At first I thought it was a missed opportunity to drop some usage of stuff in `sage.misc.package` in doctest control and then to my horror, I noticed that it created new functionality in it and spread its tentacles in an area that was blissfully free of it before: "features". In fact, I took features as way to do things without `sage.misc.packages`. Well, no more!
> 
> `sage.misc.packages` is clearly an interface to consult the sage distribution packaging system. Its inclusion, use and any extensions goes against the effort of making sage more modular and pip installable as far as I am concerned. sage should be able to run independently of sage the distribution. That's the number one thing I would like to see split as a completely separate package and stopped being relied on for testing or anything else. This PR certainly hurt sage-on-distro as far as I am concerned.

I'm not sure I understand your point correctly. Is it about this `from sage.misc.package import _spkg_type` statement in `src/sage/features/__init__.py`? This is only used to classify the given feature in terms of the type of corresponding `SPGK`. If you know a way to achieve this without the import I'd appreciate your suggestion. Of course, one possibility would be code duplication. For this, however, AFAIK files under `sage.env.SAGE_PKGS` would also have to be read.

**Comment by mkoeppe** [comment]
We could of course declare what features are standard/optional explicitly when we define a feature.
In downstream distribution packaging, `build/pkgs` is usually not available.

Of course, this is a minor point because referring to the features by group is only a convenience for developers.

**Comment by mkoeppe** [comment]
In any case, this is something that we can address in a follow up related to https://github.com/sagemath/sage/issues/35790 and https://github.com/sagemath/sage/issues/35790

**Comment by kiwifb** [comment]
I am against adding any more functionality to `sage.misc.package`, I am against adding any more calls to it than there are now. What other python package queries a specific package manager to figure if something is here? Or a classification that is only available in one specific deployment case?

I'll posit that `sage.misc.package` should be refactored out of sage altogether and never be used to figure if something is installed by the official runtime. I'll have to rip all of this ticket out of sage-on-gentoo. My fault for not having ripped `sage.misc.package` out of the doctesting framework altogether a long time ago so people would not get any more ideas. And `features` was to be our way out of this mess and now it is polluted.

**Comment by mkoeppe** [comment]
> `sage.misc.package` should [...] never be used to figure if something is installed

This PR does not use sage.misc.package for that. It only uses it to query what packages are considered standard and which are considered optional.

**Comment by mkoeppe** [comment]
But I'm working on a follow-up that you'll like.

**Pull request merged**